File: IPCalc.cs
// IPCalc.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

class IPCalc
{
    static uint IpToInt(string ip)
    {
        var parts = ip.Split('.');
        return (uint.Parse(parts[0]) << 24) |
               (uint.Parse(parts[1]) << 16) |
               (uint.Parse(parts[2]) << 8) |
               uint.Parse(parts[3]);
    }

    static string IntToIp(uint num)
    {
        return $"{((num >> 24) & 0xff)}.{((num >> 16) & 0xff)}.{((num >> 8) & 0xff)}.{num & 0xff}";
    }

    static uint PrefixToMask(int prefix) => (~0u << (32 - prefix)) & 0xffffffff;

    static int MaskToPrefix(uint mask) => System.Numerics.BitOperations.PopCount(mask);

    static string IpBinary(uint num)
    {
        return $"{Convert.ToString((num >> 24) & 0xff, 2).PadLeft(8, '0')}." +
               $"{Convert.ToString((num >> 16) & 0xff, 2).PadLeft(8, '0')}." +
               $"{Convert.ToString((num >> 8) & 0xff, 2).PadLeft(8, '0')}." +
               $"{Convert.ToString(num & 0xff, 2).PadLeft(8, '0')}";
    }

    class Result
    {
        public string ip { get; set; }
        public string mask { get; set; }
        public int prefix { get; set; }
        public string network { get; set; }
        public string broadcast { get; set; }
        public string wildcard { get; set; }
        public string first_usable { get; set; }
        public string last_usable { get; set; }
        public int total_hosts { get; set; }
        public int usable_hosts { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string binary_ip { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string binary_mask { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string binary_network { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public List<Subnet> subnets { get; set; }
    }

    class Subnet
    {
        public string network { get; set; }
        public string prefix { get; set; }
        public string first_usable { get; set; }
        public string last_usable { get; set; }
    }

    static Result Calculate(string input, string mask)
    {
        uint ipInt, maskInt;
        int prefix;
        if (input.Contains("/"))
        {
            var parts = input.Split('/');
            ipInt = IpToInt(parts[0]);
            prefix = int.Parse(parts[1]);
            maskInt = PrefixToMask(prefix);
        }
        else if (mask != null)
        {
            ipInt = IpToInt(input);
            if (mask.Contains("."))
            {
                maskInt = IpToInt(mask);
                prefix = MaskToPrefix(maskInt);
            }
            else
            {
                prefix = int.Parse(mask);
                maskInt = PrefixToMask(prefix);
            }
        }
        else throw new Exception("Invalid input");
        uint networkInt = ipInt & maskInt;
        uint broadcast = networkInt | (~maskInt & 0xffffffff);
        uint wildcard = ~maskInt & 0xffffffff;
        int total = 1 << (32 - prefix);
        int usable = Math.Max(0, total - 2);
        string first = networkInt == broadcast ? null : IntToIp(networkInt + 1);
        string last = networkInt == broadcast ? null : IntToIp(broadcast - 1);
        return new Result
        {
            ip = IntToIp(ipInt),
            mask = IntToIp(maskInt),
            prefix = prefix,
            network = IntToIp(networkInt),
            broadcast = IntToIp(broadcast),
            wildcard = IntToIp(wildcard),
            first_usable = first,
            last_usable = last,
            total_hosts = total,
            usable_hosts = usable
        };
    }

    static List<Subnet> GetSubnets(string input, string mask, int newPrefix)
    {
        var baseRes = Calculate(input, mask);
        if (newPrefix <= baseRes.prefix) throw new Exception("New prefix must be larger than original");
        uint netInt = IpToInt(baseRes.network);
        int num = 1 << (newPrefix - baseRes.prefix);
        uint step = 1u << (32 - newPrefix);
        var list = new List<Subnet>();
        for (int i = 0; i < num; i++)
        {
            uint n = netInt + (uint)i * step;
            var sub = Calculate(IntToIp(n), newPrefix.ToString());
            list.Add(new Subnet
            {
                network = sub.network,
                prefix = sub.prefix.ToString(),
                first_usable = sub.first_usable ?? "none",
                last_usable = sub.last_usable ?? "none"
            });
        }
        return list;
    }

    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: IPCalc <ip>[/prefix] [mask] [--subnet PREFIX] [--bin] [--json]");
            return;
        }
        string input = args[0];
        string mask = null;
        bool bin = false, json = false;
        int subnetPrefix = -1;
        for (int i = 1; i < args.Length; i++)
        {
            if (args[i] == "--subnet" && i + 1 < args.Length)
                subnetPrefix = int.Parse(args[++i]);
            else if (args[i] == "--bin") bin = true;
            else if (args[i] == "--json") json = true;
            else if (!args[i].StartsWith("--") && mask == null)
                mask = args[i];
        }
        var res = Calculate(input, mask);
        if (bin)
        {
            res.binary_ip = IpBinary(IpToInt(res.ip));
            res.binary_mask = IpBinary(IpToInt(res.mask));
            res.binary_network = IpBinary(IpToInt(res.network));
        }
        if (subnetPrefix != -1)
        {
            res.subnets = GetSubnets(input, mask, subnetPrefix);
        }
        if (json)
        {
            var opt = new JsonSerializerOptions { WriteIndented = true };
            Console.WriteLine(JsonSerializer.Serialize(res, opt));
            return;
        }
        Console.WriteLine("\nIP Calculator");
        Console.WriteLine($"IP Address:     {res.ip}");
        Console.WriteLine($"Subnet Mask:    {res.mask} (/{res.prefix})");
        Console.WriteLine($"Network:        {res.network}");
        Console.WriteLine($"Broadcast:      {res.broadcast}");
        Console.WriteLine($"Wildcard Mask:  {res.wildcard}");
        if (res.first_usable != null)
        {
            Console.WriteLine($"First Usable:   {res.first_usable}");
            Console.WriteLine($"Last Usable:    {res.last_usable}");
        }
        else
        {
            Console.WriteLine("First Usable:   (none)");
            Console.WriteLine("Last Usable:    (none)");
        }
        Console.WriteLine($"Total Hosts:    {res.total_hosts}");
        Console.WriteLine($"Usable Hosts:   {res.usable_hosts}");
        if (bin)
        {
            Console.WriteLine("\nBinary:");
            Console.WriteLine($"IP:      {res.binary_ip}");
            Console.WriteLine($"Mask:    {res.binary_mask}");
            Console.WriteLine($"Network: {res.binary_network}");
        }
        if (res.subnets != null && res.subnets.Count > 0)
        {
            Console.WriteLine($"\nSubnets (/{subnetPrefix}):");
            foreach (var s in res.subnets)
                Console.WriteLine($"  {s.network}/{s.prefix}   ({s.first_usable} - {s.last_usable})");
        }
    }
}

File: IPCalc.java
// IPCalc.java
import java.io.*;
import java.util.*;
import com.google.gson.*;

public class IPCalc {
    public static long ipToInt(String ip) {
        String[] parts = ip.split("\\.");
        return (Long.parseLong(parts[0]) << 24) |
               (Long.parseLong(parts[1]) << 16) |
               (Long.parseLong(parts[2]) << 8) |
               Long.parseLong(parts[3]);
    }

    public static String intToIP(long num) {
        return ((num >> 24) & 0xff) + "." +
               ((num >> 16) & 0xff) + "." +
               ((num >> 8) & 0xff) + "." +
               (num & 0xff);
    }

    public static long prefixToMask(int prefix) {
        return (~0L << (32 - prefix)) & 0xffffffffL;
    }

    public static int maskToPrefix(long mask) {
        return Long.bitCount(mask);
    }

    public static String ipBinary(long num) {
        return String.format("%08d.%08d.%08d.%08d",
            Long.parseLong(Long.toBinaryString((num >> 24) & 0xff)),
            Long.parseLong(Long.toBinaryString((num >> 16) & 0xff)),
            Long.parseLong(Long.toBinaryString((num >> 8) & 0xff)),
            Long.parseLong(Long.toBinaryString(num & 0xff)));
    }

    static class CalcResult {
        String ip, mask, network, broadcast, wildcard, firstUsable, lastUsable;
        int prefix, totalHosts, usableHosts;
    }

    public static CalcResult calculate(String input, String maskStr) throws Exception {
        long ipInt, maskInt;
        int prefix;
        if (input.contains("/")) {
            String[] parts = input.split("/");
            ipInt = ipToInt(parts[0]);
            prefix = Integer.parseInt(parts[1]);
            maskInt = prefixToMask(prefix);
        } else if (maskStr != null) {
            ipInt = ipToInt(input);
            if (maskStr.contains(".")) {
                maskInt = ipToInt(maskStr);
                prefix = maskToPrefix(maskInt);
            } else {
                prefix = Integer.parseInt(maskStr);
                maskInt = prefixToMask(prefix);
            }
        } else {
            throw new Exception("Invalid input");
        }
        long networkInt = ipInt & maskInt;
        long broadcast = networkInt | (~maskInt & 0xffffffffL);
        long wildcard = ~maskInt & 0xffffffffL;
        int total = 1 << (32 - prefix);
        int usable = Math.max(0, total - 2);
        long firstUsable = networkInt == broadcast ? -1 : networkInt + 1;
        long lastUsable = networkInt == broadcast ? -1 : broadcast - 1;

        CalcResult res = new CalcResult();
        res.ip = intToIP(ipInt);
        res.mask = intToIP(maskInt);
        res.prefix = prefix;
        res.network = intToIP(networkInt);
        res.broadcast = intToIP(broadcast);
        res.wildcard = intToIP(wildcard);
        res.firstUsable = firstUsable != -1 ? intToIP(firstUsable) : null;
        res.lastUsable = lastUsable != -1 ? intToIP(lastUsable) : null;
        res.totalHosts = total;
        res.usableHosts = usable;
        return res;
    }

    public static List<Map<String, String>> subnets(String input, String maskStr, int newPrefix) throws Exception {
        CalcResult base = calculate(input, maskStr);
        if (newPrefix <= base.prefix) throw new Exception("New prefix must be larger than original");
        long netInt = ipToInt(base.network);
        int num = 1 << (newPrefix - base.prefix);
        long step = 1L << (32 - newPrefix);
        List<Map<String, String>> result = new ArrayList<>();
        for (int i = 0; i < num; i++) {
            long n = netInt + i * step;
            CalcResult sub = calculate(intToIP(n), String.valueOf(newPrefix));
            Map<String, String> map = new HashMap<>();
            map.put("network", sub.network);
            map.put("prefix", String.valueOf(sub.prefix));
            map.put("first_usable", sub.firstUsable != null ? sub.firstUsable : "none");
            map.put("last_usable", sub.lastUsable != null ? sub.lastUsable : "none");
            result.add(map);
        }
        return result;
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("Usage: IPCalc <ip>[/prefix] [mask] [--subnet PREFIX] [--bin] [--json]");
            return;
        }
        String input = args[0];
        String mask = args.length > 1 && !args[1].startsWith("--") ? args[1] : null;
        boolean bin = false, json = false;
        int subnetPrefix = -1;
        for (int i = 1; i < args.length; i++) {
            if (args[i].equals("--subnet") && i+1 < args.length) {
                subnetPrefix = Integer.parseInt(args[++i]);
            } else if (args[i].equals("--bin")) {
                bin = true;
            } else if (args[i].equals("--json")) {
                json = true;
            }
        }
        CalcResult res = calculate(input, mask);
        Map<String, Object> resultMap = new LinkedHashMap<>();
        resultMap.put("ip", res.ip);
        resultMap.put("mask", res.mask);
        resultMap.put("prefix", res.prefix);
        resultMap.put("network", res.network);
        resultMap.put("broadcast", res.broadcast);
        resultMap.put("wildcard", res.wildcard);
        resultMap.put("first_usable", res.firstUsable);
        resultMap.put("last_usable", res.lastUsable);
        resultMap.put("total_hosts", res.totalHosts);
        resultMap.put("usable_hosts", res.usableHosts);
        if (bin) {
            resultMap.put("binary_ip", ipBinary(ipToInt(res.ip)));
            resultMap.put("binary_mask", ipBinary(ipToInt(res.mask)));
            resultMap.put("binary_network", ipBinary(ipToInt(res.network)));
        }
        List<Map<String, String>> subnets = null;
        if (subnetPrefix != -1) {
            subnets = subnets(input, mask, subnetPrefix);
            resultMap.put("subnets", subnets);
        }

        if (json) {
            Gson gson = new GsonBuilder().setPrettyPrinting().create();
            System.out.println(gson.toJson(resultMap));
            return;
        }
        System.out.println("\nIP Calculator");
        System.out.printf("IP Address:     %s%n", res.ip);
        System.out.printf("Subnet Mask:    %s (/%d)%n", res.mask, res.prefix);
        System.out.printf("Network:        %s%n", res.network);
        System.out.printf("Broadcast:      %s%n", res.broadcast);
        System.out.printf("Wildcard Mask:  %s%n", res.wildcard);
        if (res.firstUsable != null) {
            System.out.printf("First Usable:   %s%n", res.firstUsable);
            System.out.printf("Last Usable:    %s%n", res.lastUsable);
        } else {
            System.out.println("First Usable:   (none)");
            System.out.println("Last Usable:    (none)");
        }
        System.out.printf("Total Hosts:    %d%n", res.totalHosts);
        System.out.printf("Usable Hosts:   %d%n", res.usableHosts);
        if (bin) {
            System.out.println("\nBinary:");
            System.out.printf("IP:      %s%n", resultMap.get("binary_ip"));
            System.out.printf("Mask:    %s%n", resultMap.get("binary_mask"));
            System.out.printf("Network: %s%n", resultMap.get("binary_network"));
        }
        if (subnets != null && !subnets.isEmpty()) {
            System.out.printf("\nSubnets (/%d):%n", subnetPrefix);
            for (Map<String, String> s : subnets) {
                System.out.printf("  %s/%s   (%s - %s)%n", s.get("network"), s.get("prefix"), s.get("first_usable"), s.get("last_usable"));
            }
        }
    }
}
